### DISCUSSION TITLE
Usage of nginx-proxy in a production environment

### DISCUSSION THREAD

**Discussion by 0xclx** [nginx-proxy/nginx-proxy#1633] [Q&A]
Is nginx-proxy suitable for use in a production environment? It doesn't use the privileged mode, but it mounts the docker-socket into the container which apparently exposes a trivial way to escalate privileges and gain access to the host.

So, is nginx-proxy a suitable way to dynamically add websites on subdomains in a production environment?
If not, what would be the best practice alternative? Manually configure a NGINX-Server hosted on the server itself?

I keep finding conflicting information about how to use Nginx in Production - And would like to come up with a clear, recommended way.

**Comment by kerberjg** [comment]
A solution to the docker socket issue you mentioned, is to use the [split-container configuration](https://github.com/nginx-proxy/nginx-proxy#separate-containers). This way, there's no way for that exploit to be performed

Otherwise, I wouldn't know how to answer your question. I have been successfully using nginx-proxy, but it a non-critical application (simple frontend). I can assure you it's stable in that, but there hasn't been much development

**Comment by LarsMS97** [comment]
But the split container configuration also mounts the docker socket. Its right in the second line:
```sh
$ docker run --volumes-from nginx \
    -v /var/run/docker.sock:/tmp/docker.sock:ro \
    -v $(pwd):/etc/docker-gen/templates \
    -t jwilder/docker-gen -notify-sighup nginx -watch /etc/docker-gen/templates/nginx.tmpl /etc/nginx/conf.d/default.conf
```

 How would you use this on e.g. azure, where accessing the docker socket is [not possible](https://stackoverflow.com/questions/49977281/how-to-access-docker-sock-on-a-container-instance-inside-an-azure-container-grou)?

**Comment by 0x47** [comment]
I think it's a huge red flag that this project does not explicitly state that the current recommended setup is not secure at all. The only mention of this is in [this sentence](https://github.com/nginx-proxy/nginx-proxy/blame/master/README.md#L153):

> You may want to do this to prevent having the docker socket bound to a publicly exposed container service.

Failing to communicate this issue has led people to suggesting this project as convenient way to set up production environment servers, e.g., [Nextcloud](https://blog.ssdnodes.com/blog/installing-nextcloud-docker/). I was also looking for a nice way to set up a Nextcloud server via docker-compose but the mounting of the docker socket has led me here to this GitHub issue.

[Here](https://raesene.github.io/blog/2016/03/06/The-Dangers-Of-Docker.sock/) is one reason why it is a bad idea to do this at all:

> [...] actually the ro setting on the volume mount doesn’t have a lot of effect in terms of security. An attacker with ro access to the socket can still create another container and do something like mount /etc/ into it from the host, essentially giving them root access to the host.

I think this problem should either be made _very_ clear in the README instructions or the default setup should use the split container configuration. Or maybe, might there be a clever way to implement nginx-proxy while not mounting the socket at all?

EDIT: On that last note: Maybe [this project](https://github.com/Tecnativa/docker-socket-proxy) could be integrated per default until a better solution is found.
